tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala

```diff
@@ -1,13 +1,99 @@
+import uk.ac.wellcome.json.JsonUtil._
+import uk.ac.wellcome.messaging.MessageSender
+import uk.ac.wellcome.messaging.sns.NotificationMessage
 import uk.ac.wellcome.messaging.sqs.SQSStream
+import uk.ac.wellcome.storage.s3.S3ObjectLocation
+import uk.ac.wellcome.storage.store.Writable
 import uk.ac.wellcome.typesafe.Runnable
+import weco.flows.FlowOps
 
-import scala.concurrent.Future
+import java.net.URI
+import java.time.Instant
+import scala.concurrent.{ExecutionContext, Future}
 
-case class TeiIdExtractorWorkerService(messageStream: SQSStream[Nothing],
-                                       messageSender: SNSMessageSender)
-    extends Runnable {
-  override def run(): Future[Any] = ???
+// Represents a path change coming from the tei_updater lambda
+case class TeiFileChangedMessage(path: String, uri: URI, timeModified: Instant)
+
+// Represents a message for the tei_adapter with changes to id instead of file path
+case class TeiIdChangeMessage(id: String,
+                              s3Location: S3ObjectLocation,
+                              timeModified: Instant)
+case class TeiIdExtractorConfig(concurrentFiles: Int, bucket: String)
+
+class TeiIdExtractorWorkerService[Dest](
+  messageStream: SQSStream[NotificationMessage],
+  gitHubBlobReader: GitHubBlobContentReader,
+  store: Writable[S3ObjectLocation, String],
+  messageSender: MessageSender[Dest],
+  config: TeiIdExtractorConfig)(implicit val ec: ExecutionContext)
+    extends Runnable
+    with FlowOps {
+  val className = this.getClass.getSimpleName
+
+  override def run(): Future[Any] = {
+    messageStream.runStream(
+      className,
+      source =>
+        source
+          .via(unwrapMessage)
+          .via(broadcastAndMerge(filterNonTei, processMessage))
+          .map { case (Context(msg), _) => msg })
+  }
+
+  def unwrapMessage =
+    Flow[(SQSMessage, NotificationMessage)]
+      .map {
+        case (msg, NotificationMessage(body)) =>
+          (Context(msg), fromJson[TeiFileChangedMessage](body).toEither)
+      }
+      .via(catchErrors)
+
+  def filterNonTei = Flow[(Context, TeiFileChangedMessage)].filter {
+    case (_, TeiFileChangedMessage(path, _, _)) => !isTeiFile(path)
+  }
+
+  def processMessage =
+    Flow[(Context, TeiFileChangedMessage)]
+      .filter {
+        case (_, TeiFileChangedMessage(path, _, _)) => isTeiFile(path)
+      }
+      .mapAsync(config.concurrentFiles) {
+        case (ctx, message) =>
+          for {
+            blobContent <- gitHubBlobReader.getBlob(message.uri)
+            id <- Future.fromTry(
+              IdExtractor.extractId(blobContent, message.uri))
+            stored <- Future.fromTry(
+              store
+                .put(S3ObjectLocation(
+                  config.bucket,
+                  s"tei_files/$id/${message.timeModified.getEpochSecond}.xml"))(
+                  blobContent)
+                .left
+                .map(error => error.e)
+                .toTry)
```

Comment:
Break this up to make it more readable:

```suggestion
            location = S3ObjectLocation(
              bucket = config.bucket,
              key = s"tei_files/$id/${message.timeModified.getEpochSecond}.xml")    
            )
            stored <- store.put(location, blobContent) match {
              case Right(stored) => Future.successful(stored)
              case Failure(error) => Future.failed(error.e)
            }
```